blockdag/blocklocator.go

```diff
@@ -70,51 +62,49 @@ func (dag *BlockDAG) blockLocator(startNode, stopNode *blockNode) BlockLocator {
 	// block locator won't contain the start node.
 	startNode = startNode.selectedParent
 
-	// If the start node or the stop node are not in the
-	// virtual's selected parent chain, we replace them with their
-	// closest selected parent that is part of the virtual's
-	// selected parent chain.
-	for !dag.IsInSelectedParentChain(stopNode.hash) {
-		stopNode = stopNode.selectedParent
-	}
-
-	for !dag.IsInSelectedParentChain(startNode.hash) {
-		startNode = startNode.selectedParent
-	}
-
-	// Calculate the max number of entries that will ultimately be in the
+	// Calculate the max number of entries that could possibly be in the
 	// block locator. See the description of the algorithm for how these
 	// numbers are derived.
 
 	// startNode.hash + stopNode.hash.
-	// Then floor(log2(startNode.chainHeight-stopNode.chainHeight)) entries for the skip portion.
-	maxEntries := 2 + util.FastLog2Floor(startNode.chainHeight-stopNode.chainHeight)
+	// Then floor(log2(K*(startNode.blueScore-stopNode.blueScore))) entries for the skip portion.
+	maxEntries := 2 + util.FastLog2Floor(uint64(dag.dagParams.K)*(startNode.blueScore-stopNode.blueScore))
 	locator := make(BlockLocator, 0, maxEntries)
 
+	node := startNode
 	step := uint64(1)
-	for node := startNode; node != nil; {
+	for node != nil {
 		locator = append(locator, node.hash)
 
+		if node.blueScore < stopNode.blueScore {
+			return nil, errors.Errorf("startNode and stopNode are not " +
+				"in the same selected parent chain.")
+		}
+
 		// Nothing more to add once the stop node has been added.
-		if node.chainHeight == stopNode.chainHeight {
+		if node.blueScore == stopNode.blueScore {
+			if node.hash != stopNode.hash {
+				return nil, errors.Errorf("startNode and stopNode are " +
+					"not in the same selected parent chain.")
+			}
 			break
 		}
 
-		// Calculate chainHeight of previous node to include ensuring the
+		// Calculate blueScore of previous node to include ensuring the
 		// final node is stopNode.
-		nextChainHeight := node.chainHeight - step
-		if nextChainHeight < stopNode.chainHeight {
-			nextChainHeight = stopNode.chainHeight
+		nextBlueScore := node.blueScore - step
+		if nextBlueScore < stopNode.blueScore {
+			nextBlueScore = stopNode.blueScore
 		}
 
 		// walk backwards through the nodes to the correct ancestor.
-		node = node.SelectedAncestor(nextChainHeight)
+		node = node.SelectedAncestor(nextBlueScore)
```

Comment:
Do you really need to just to selectedAncestor using blueScores?

Wouldn't it make more sense to just jump the number of steps required?